openstack_terraforming/compute.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2018 The Terraformer Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package openstack_terraforming
+
+import (
+	"strings"
+
+	"github.com/GoogleCloudPlatform/terraformer/terraform_utils"
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+	"github.com/gophercloud/gophercloud/pagination"
+)
+
+type ComputeGenerator struct {
+	OpenStackService
+}
+
+// createResources iterate on all openstack_compute_instance_v2
+func (g *ComputeGenerator) createResources(list *pagination.Pager) []terraform_utils.Resource {
+	resources := []terraform_utils.Resource{}
+
+	list.EachPage(func(page pagination.Page) (bool, error) {
+		servers, err := servers.ExtractServers(page)
+		if err != nil {
+			return false, err
+		}
+
+		for _, s := range servers {
+			resource := terraform_utils.NewResource(
+				s.ID,
+				s.Name,
+				"openstack_compute_instance_v2",
+				"openstack",
+				map[string]string{
+					"name": s.Name,
+				},
+				[]string{},
+				map[string]string{},
+			)
+
+			resources = append(resources, resource)
+		}
+
+		return true, nil
+	})
+
+	return resources
+}
+
+// Generate TerraformResources from OpenStack API,
+func (g *ComputeGenerator) InitResources() error {
+	opts, err := openstack.AuthOptionsFromEnv()
+	if err != nil {
+		return err
+	}
+
+	provider, err := openstack.AuthenticatedClient(opts)
+	if err != nil {
+		return err
+	}
+
+	client, err := openstack.NewComputeV2(provider, gophercloud.EndpointOpts{
+		Region: g.GetArgs()["region"],
+	})
+	if err != nil {
+		return err
+	}
+
+	list := servers.List(client, nil)
+
+	g.Resources = g.createResources(&list)
+	g.PopulateIgnoreKeys()
+
+	return nil
+}
+
+func (g *ComputeGenerator) PostConvertHook() error {
+	for i, r := range g.Resources {
+		if r.InstanceInfo.Type != "openstack_compute_instance_v2" {
+			continue
+		}
+
+		// Rename "all_metadata" to "metadata"
+		// because "all_metadata" field cannot be set as resource argument
+		for k, v := range g.Resources[i].InstanceState.Attributes {
```

Comment:
You tested this? InstanceState.Attributes it same data from provider and from Attributes terraformer create tfstate.
I don't know if you need replace in tftstate "all_metadata" => "metadata". I see in tf this metadata, but I think in tfstate this all_metadata
https://github.com/terraform-providers/terraform-provider-openstack/blob/master/openstack/resource_openstack_compute_instance_v2.go#L597

Reply:
I have reconfirm the exported files of terraform-provider-openstack. I created a `tf` file containing "openstack_compute_instance_v2" and "metadata" field. As a result of `terraform apply`, both "metadata" and "all_metadata" fields were recorded in `terraform.tfstate`.
```json
            "resources": {
                "openstack_compute_instance_v2.test01": {
                    "type": "openstack_compute_instance_v2",
                    "depends_on": [],
                    "primary": {
                        "id": "f29e21b5-47b6-40b1-ae78-cd4993bd39aa",
                        "attributes": {
                            ...
                            "all_metadata.%": "2",
                            "all_metadata.foo": "bar",
                            "all_metadata.terraformer": "test",
                            ...
                            "metadata.%": "2",
                            "metadata.foo": "bar",
                            "metadata.terraformer": "test",
```

I will fix the code to record both fields to `tfstate`.
Replacing field name is still necessary for `tf` files, because "openstack_compute_instance_v2" resource does not supports "all_metadata" field.
https://www.terraform.io/docs/providers/openstack/r/compute_instance_v2.html#argument-reference